woocommerce/woocommerce-gutenberg-products-block
# Add CI for monitoring TypeScript errors

This PR adds an action for monitoring the TypeScript errors in our codebase.

Fixes #6068

The action use `typescript-checkstyle` package that generates a valid [checkstyle file](https://checkstyle.sourceforge.io/) that contains all the errors and the relative files.

After a developer creates a PR, the action will:
- generate a new `checkstyle.xml`, and if there are changes, automatically will push it into the branch.
- add a comment on Github with the number of errors and the number of files that contain errors.  
- check the generated `checkstyle.xml` with the trunk version and compare the two files: if the report of the current branch has more errors than the trunk version, the action will update the comment and will add the path of the files that have the new errors.

Also, at each push in the trunk (every time that a PR is merged in the trunk) the action will update a table on AirTable (not publicly visible), where a linked chart shows the migration status.

## Note 

At this stage, the diff between the `checkstyle` in the current branch and the `checkstyle` in the trunk branch is handled by checking the number of errors for each file. So, for instance, on a specific file if the developer fixes an error, but introduces another one the current implementation will not catch the error, so the action will not add any warning in the comment.


I removed the nested `tsconfig.json` configurations because they did not extend any rules and tsc gave me a lot of errors as this:
```
File 'xxx/xxx/xxx.ts' is not listed within the file list of project 'xxx/xxx/tsconfig.json'. Projects must list all files or use an 'include' pattern.
```


### Testing instructions

Now, the action is failing because the branch to compare is the trunk (that doesn't have a `checkstyle.xml`). I prepared a branch for testing that is equal to this branch.

Check out the branch and update the code following the next instructions.
It is necessary to update [these line](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/19264a4bcf3bb2f210883c3278fd9c5374e55ae7/.github/monitor-typescript-errors/utils/github.js/#L9-L13) and replace with:
``` javascript
return await octokit.rest.repos.getContent( {
    owner,
    repo,
    ref: "try/typescript-monitor-second-branch"
    path: fileName,
    mediaType: {
      format: 'raw',
      },
} );
```

Please, remember that you need to create a PR to trigger the action.

You should check that:

### Action pushes automatically a new checkstyle.xml is something in the codebase change

You can add a new TS error in any file in the codebase. The action should be to push the updated checkstyle automatically.

### Action adds a warning if you introduce a new error in the codebase

You can add a new TS error in any file in the codebase. The actions should add a comment with a list of the file that has new errors.

### Action updates the AirTable table

For this, you need update [this check](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/6106/files#diff-0142a71a92636a1f9867130ef62411824f63a59f3adcaaaa4d2625464edccf96R56) and replace `trunk` with try/typescript-monitor-second-branch.

If you visit our AirTable and click on `TypeScript migration`, you should see that there is a new row.




## NOTE

Now, the Monitor TypeScript errors fail in this branch because the job compares the `checkstyle.xml` with the trunk branch, but at this stage, the `checkstyle.xml` is missing in the trunk. This is the reason it is necessary to tweak the file `github.js`.
